runtime/compiler/optimizer/UnsafeFastPath.cpp

```diff
@@ -279,6 +280,26 @@ bool TR_UnsafeFastPath::tryTransformUnsafeAtomicCallInVarHandleAccessMethod(TR::
       if (isVarHandleOperationMethodOnNonStaticField(callerMethod) &&
          performTransformation(comp(), "%s transforming Unsafe.CAS [" POINTER_PRINTF_FORMAT "] into codegen inlineable\n", optDetailString(), node))
          {
+      #if defined(J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+         if (isUnsafeCallerAccessingArrayElement(callerMethod) && TR::Compiler->om.isOffHeapAllocationEnabled() && comp()->target().is64Bit())
```

Comment:
this appears to assume: at compile time, it is definitely known if it is array access or not?